solutions/cf-1017G.mdx

```diff
@@ -0,0 +1,218 @@
+---
+id: cf-1017G
+source: CF
+title: The Tree
+author: Dustin Miao
+---
+
+## Solution
+
+The [official editorial](https://codeforces.com/blog/entry/61081) mentions  using [Heavy-Light Decomposition](https://usaco.guide/plat/hld) to solve this problem, but never discusses the solution. 
+
+In each node, store a **propagation factor**. This is essentially how much black has spread from the node, or equivalently, the number of operations of type 1 that has been performed. White nodes have a propagation factor of $-1$ and black nodes have a propagation factor that is greater than or equal to zero. An operation of type 1 now directly corresponds with increasing the propagation factor of a node by $1$. 
+
+Now consider operation 3 (and ignore operation 2 for now). How do we determine if a given node is black? Firstly, note that if the current node has a non-negative propagation factor, then it must be black. If instead its parent had a propagation factor greater than or equal to one, then the node would also be black. 
+
+In general, consider if operation 3 was called on some node $u$. Let $v$ be any node that lies on the path from $u$ to the root. If the sum of propagation factors from $u$ to $v$ is non-negative, then $u$ must be black. What this means is that enough operations of type 1 have been applied from that node (and possibly other nodes on the path from $u$ to $v$) such that $u$ has been made black. 
+
+For operations of type 2, we can easily clear a given subtree by taking advantage of the fact that HLD is really just a specific euler tour. Thus, a subtree of a node lies in a contiguous range in the heavy-light decomposition. We can clear a subtree by using lazy propagation and setting all nodes in a range to $-1$. Next, we have to adjust the propagation factors of all nodes above $u$, where the operation was performed, which may take up to linear time. A simple trick is to subtract a factor of $q - 1$ from node $u$, where $q$ is the value of a operation of type 3. This ensures that any query within $u$'s subtree will correctly evaluate to white after the update. 
+
+Note that for query operations, we need to find the maximum suffix sum for each path. This can be done by maintaing the maximum suffix sum as well as the actual sum in each segment tree node. 
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(n + q\log^2n)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+const int MAXN = 100001;
+
+using ll = long long
+using pll = pair<ll, ll>;
+const ll INF = LLONG_MAX;
+#define FF first
+#define SS second
+
+int N, Q;
+vector<int> G[MAXN];
+
+namespace sgt {
```

Comment:
comment explaining what the segtree & hld thing are used for

Reply:
...well of course it's a segment tree, but what is it _USED_ for?
"segment tree for keeping track of [asdf]"

Reply:
it's explained in the solution section

Reply:
also hld typically uses segtree as a subroutine
it wouldn't not make sense if the sgt namespace was inside the hld namespace

Reply:
> it's explained in the solution section

no harm in explaining it again
think it's common to put small comments so people can link the stuff in the editorial to the actual code

Reply:
> it wouldn't not make sense

read this again

Reply:
Double negative makes sense in this context
sort of like "Don't say I didn't warn you" vs "I warned you"

Reply:
oh you meant that it _would_ make sense if the namespace was nested
i thought you were meaning to say that it wouldn't